### PR TITLE
cmake: Switch from tri-state options to boolean. Stage ONE

### PR DESCRIPTION
**Pull request by hebasto**
This PR partially pulled from https://github.com/hebasto/bitcoin/pull/83 with options, which default values seem non-controversial.

As a bonus, the help string duplication in `depends/toolchain.cmake.in` has been eliminated for the touched lines.

The  [**Autotools vs CMake Feature Parity Table**](https://gist.github.com/hebasto/2ef97d3a726bfce08ded9df07f7dab5e) has been updated as well.

### CODE REVIEW

File: depends/toolchain.cmake.in [hunk 1]
```diff
@@ -138,27 +140,31 @@ if(NOT WITH_SQLITE AND "@no_sqlite@" STREQUAL "1")
...
-    set(MULTIPROCESS ON CACHE STRING "")
-  endif()
-  if(NOT LibmultiprocessGen_DIR)
-    set(LibmultiprocessGen_DIR "${CMAKE_FIND_ROOT_PATH}/native/lib/cmake/LibmultiprocessGen" CACHE PATH "")
-  endif()
+  set(MULTIPROCESS ON CACHE BOOL "")
+  set(LibmultiprocessGen_DIR "${CMAKE_FIND_ROOT_PATH}/native/lib/cmake/LibmultiprocessGen" CACHE PATH "")
+else()
+  set(MULTIPROCESS OFF CACHE BOOL "")
 endif()
```
Comment (vasild) [inline depends/toolchain.cmake.in:170]: `MULTIPROCESS` is a boolean like e.g. `WITH_NATPMP` but is handled differently here. Why? Should this not be:

```cmake
if("@no_multiprocess@")
  set(MULTIPROCESS OFF CACHE BOOL "")
else()
  set(MULTIPROCESS ON CACHE BOOL "")
  set(LibmultiprocessGen_DIR "${CMAKE_FIND_ROOT_PATH}/native/lib/cmake/LibmultiprocessGen" CACHE PATH "")
endif()
```
?

File: depends/toolchain.cmake.in [hunk 2]
```diff
@@ -138,27 +140,31 @@ if(NOT WITH_SQLITE AND "@no_sqlite@" STREQUAL "1")
   set(WITH_SQLITE OFF CACHE STRING "Enable SQLite wallet support.")
 endif()
 
-if(NOT WITH_MINIUPNPC AND "@no_upnp@" STREQUAL "1")
-  set(WITH_MINIUPNPC OFF CACHE STRING "Enable UPnP.")
+if("@no_upnp@")
+  set(WITH_MINIUPNPC OFF CACHE BOOL "")
```
Comment (vasild) [inline depends/toolchain.cmake.in:144]: I am not sure how those `"@no_foo@"` are generated, but there is some inconsistency between upnp and the others. For example, `CMakeLists.txt` defines `WITH_NATPMP` and in `depends/toolchain.cmake.in` the check is using `no_natpmp`. But for upnp: `CMakeLists.txt`: `WITH_MINIUPNPC`, `depends/toolchain.cmake.in`: `no_upnp`.

**Review by vasild — COMMENTED** [review]
Approach ACK d66f911b68

File: depends/toolchain.cmake.in (hunk 2 above)
Comment (hebasto) [reply depends/toolchain.cmake.in:144]: > I am not sure how those "@no_foo@" are generated...

https://github.com/hebasto/bitcoin/blob/d66f911b6863bb3b20e8e28eb14edceb9db0d555/depends/Makefile#L288-L298

> ... but there is some inconsistency between upnp and the others. For example, `CMakeLists.txt` defines `WITH_NATPMP` and in `depends/toolchain.cmake.in` the check is using `no_natpmp`. But for upnp: `CMakeLists.txt`: `WITH_MINIUPNPC`, `depends/toolchain.cmake.in`: `no_upnp`.

You mean, `WITH_MINIUPNPC` should be renamed to `WITH_UPNP`?

File: .github/workflows/cmake.yml [hunk 3]
```diff
@@ -227,7 +227,7 @@ jobs:
 
       - name: Generate build system
         run: |
-          cmake -B build -DCMAKE_C_COMPILER=gcc-11 -DCMAKE_CXX_COMPILER=g++-11 -DBoost_INCLUDE_DIR="${PWD}/${{ matrix.conf.boost_archive }}" -DENABLE_WALLET=ON -DWITH_EXTERNAL_SIGNER=ON -DWITH_MINIUPNPC=ON -DWITH_NATPMP=ON -DWITH_ZMQ=ON -DWITH_USDT=ON -DWERROR=ON
+          cmake -B build --preset ci-linux -DWITH_GUI=OFF -DBoost_INCLUDE_DIR="${PWD}/${{ matrix.conf.boost_archive }}"
```
[the remark this answers is not in the document]
Comment (hebasto) [reply .github/workflows/cmake.yml]: > I would expect commit [a12b692](https://github.com/hebasto/bitcoin/commit/a12b692e118f1b35cfb6994177b0a2f52215854b) `cmake: Add CI-specific presets` to just move options from the command line to the presets file, however this is not the case.

Thanks! Reworked to meet expectations.

File: depends/toolchain.cmake.in (hunk 2 above)
Comment (vasild) [reply depends/toolchain.cmake.in:144]: > You mean, `WITH_MINIUPNPC` should be renamed to `WITH_UPNP`?

Either this or change the above to `-e 's|@no_miniupnpc@|$(NO_MINIUPNPC)|'`. No strong opinion, but it is nice to be more consistent.

**Review by vasild — APPROVED** [review]
ACK df4e46bfaf2c2782d949291cc5273db64782f7cc

File: .github/workflows/cmake.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (vasild) [reply .github/workflows/cmake.yml]: Thanks! Do you think it makes sense to have all options in the presents file, so that they are not scattered between the command line and the presets file? Or would that be too un-flexible?

File: CMakeLists.txt [hunk 4]
```diff
@@ -70,9 +70,32 @@ option(REDUCE_EXPORTS "Attempt to reduce exported symbols in the resulting execu
 option(WERROR "Treat compiler warnings as errors." OFF)
 
 tristate_option(CCACHE "Use ccache for compiling." "if ccache is found." AUTO)
-tristate_option(WITH_NATPMP "Enable NAT-PMP." "if libnatpmp is found." AUTO)
-tristate_option(WITH_MINIUPNPC "Enable UPnP." "if libminiupnpc is found." AUTO)
-tristate_option(WITH_ZMQ "Enable ZMQ notifications." "if libzmq is found." AUTO)
+
+option(WITH_NATPMP "Enable NAT-PMP." OFF)
```
Comment (fanquake) [inline CMakeLists.txt:74]: Aren't we meant to be reming these, or is that happening at a later point?

File: cmake/optional.cmake [hunk 5]
```diff
@@ -189,12 +135,13 @@ else()
 endif()
 
 if(WITH_GUI AND WITH_QRENCODE)
-  if(MSVC)
+  if(VCPKG_TARGET_TRIPLET)
```
Comment (fanquake) [inline cmake/optional.cmake:138]: Why was this changed?

File: CMakeLists.txt (hunk 4 above)
Comment (hebasto) [reply CMakeLists.txt:74]: Yes, I'm going to rename all options at once at a later point.

File: CMakeLists.txt [hunk 6]
```diff
@@ -70,9 +70,32 @@ option(REDUCE_EXPORTS "Attempt to reduce exported symbols in the resulting execu
...
+endif()
+
+option(WITH_MINIUPNPC "Enable UPnP." OFF)
+if(WITH_MINIUPNPC)
+  find_package(MiniUPnPc MODULE REQUIRED)
+endif()
+
+option(WITH_ZMQ "Enable ZMQ notifications." OFF)
+if(WITH_ZMQ)
+  if(VCPKG_TARGET_TRIPLET)
```
[the remark this answers is not in the document]
Comment (hebasto) [reply CMakeLists.txt:86]: The `find_package(ZeroMQ CONFIG)` invocation is recommended by the vcpkg. I didn't test pkg-config approach for it.

File: cmake/optional.cmake (hunk 5 above)
Comment (hebasto) [reply cmake/optional.cmake:138]: This makes an intention clearer because the logic is specific to vcpkg package manager, not to the MSVC compiler.

Comment (fanquake) [reply cmake/optional.cmake:138]: Is anyone using MSVC without vcpkg? Is this being done consistently?

Comment (hebasto) [reply cmake/optional.cmake:138]: > Is anyone using MSVC without vcpkg?

I don't think so. But vcpkg might be used on other OSes.

> Is this being done consistently?

I hope it is consistent now in the staging branch.
